controlplane/pkg/nsm/nsm.go

```diff
@@ -534,7 +534,7 @@ func (srv *networkServiceManager) updateConnection(ctx context.Context, conn con
 		return conn, err
 	}
 
-	return wrapper.GetConnection(), nil
+	return wrapper.GetConnection().Clone(), nil
```

Comment:
Maybe it is not a proper place to call clone(). Seems we have to clone a ConnectionWrapper instance (not a Connection instance) in the gRPC service implementation, am I wrong?

Reply:
```
diff --git a/controlplane/pkg/apis/plugins/connection_plugin_helpers.go b/controlplane/pkg/apis/plugins/connection_plugin_helpers.go
index 0e9e50af..2ef42a5f 100644
--- a/controlplane/pkg/apis/plugins/connection_plugin_helpers.go
+++ b/controlplane/pkg/apis/plugins/connection_plugin_helpers.go
@@ -1,6 +1,8 @@
 package plugins
 
 import (
+	"github.com/gogo/protobuf/proto"
+
 	local "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
 	"github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/nsm/connection"
 	remote "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/remote/connection"
@@ -33,3 +35,8 @@ func (w *ConnectionWrapper) SetConnection(conn connection.Connection) {
 		}
 	}
 }
+
+// Clone clones wrapper
+func (w *ConnectionWrapper) Clone() *ConnectionWrapper {
+	return proto.Clone(w).(*ConnectionWrapper)
+}
diff --git a/controlplane/pkg/plugins/connection_plugin.go b/controlplane/pkg/plugins/connection_plugin.go
index 0831ace5..8c29ec5d 100644
--- a/controlplane/pkg/plugins/connection_plugin.go
+++ b/controlplane/pkg/plugins/connection_plugin.go
@@ -64,7 +64,7 @@ func (cpm *connectionPluginManager) UpdateConnection(ctx context.Context, wrappe
 			return nil, fmt.Errorf("'%s' connection plugin returned an error: %v", name, err)
 		}
 	}
-	return wrapper, nil
+	return wrapper.Clone(), nil
 }
 
 func (cpm *connectionPluginManager) ValidateConnection(ctx context.Context, wrapper *plugins.ConnectionWrapper) (*plugins.ConnectionValidationResult, error) {
```

Reply:
Or we have to clone the wrapper before we start to modify it (at the beginning of UpdateConnection() method)

Reply:
Discussed on call, decided to call Clone() in the plugin manager

Reply:
Done. Thank you!